lib/LongVectorLoweringPass.cpp

```diff
@@ -291,44 +325,201 @@ Value *LongVectorLoweringPass::visit(Value *V) {
     return it->second;
   }
 
-  if (auto *BinOp = dyn_cast<BinaryOperator>(V)) {
-    if (auto *EquivalentType = getEquivalentType(BinOp->getType())) {
-      IRBuilder<> B(BinOp);
+  if (isa<Argument>(V)) {
+    assert(getEquivalentType(V->getType()) == nullptr &&
+           "Argument not handled when visiting function.");
+    return nullptr;
+  }
 
-      SmallVector<Value *, 16> EquivalentArgs;
-      for (auto &Operand : BinOp->operands()) {
-        auto *Arg = Operand.get();
-        // TODO Visit argument to lower it.
-        // Instead, for now, we create an equivalent aggregate.
-        auto *EquivalentArgTy = getEquivalentType(Arg->getType());
-        auto *EquivalentArg = convertEquivalentValue(B, Arg, EquivalentArgTy);
-        EquivalentArgs.push_back(EquivalentArg);
-      }
+  assert(isa<User>(V) && "Kind of llvm::Value not yet supported.");
 
-      auto ScalarFactory = [Opcode = BinOp->getOpcode()](auto &B, auto Args) {
-        return B.CreateNAryOp(Opcode, Args);
-      };
-      Value *EquivalentValue = convertVectorOperation(
-          B, EquivalentType, EquivalentArgs, ScalarFactory);
+  if (auto *I = dyn_cast<Instruction>(V)) {
+    if (handlingRequired(*I)) {
+      // Dispatch to the appropriate method using InstVisitor.
+      return visit(I);
+    }
 
-      // TODO Implement support for additional instructions.
-      // Because support is very limited, as an initial step we convert the
-      // aggregate back to a vector to generate a valid module.
-      auto *ReplacementValue =
-          convertEquivalentValue(B, EquivalentValue, BinOp->getType());
-      registerReplacement(*BinOp, *ReplacementValue);
+    return nullptr;
+  }
 
-      return ReplacementValue;
-    }
+  if (auto *C = dyn_cast<Constant>(V)) {
+    return visitConstant(*C);
   }
 
 #ifndef NDEBUG
   dbgs() << "Value not handled: " << *V << '\n';
 #endif
-  // TODO Once additional features are implemented, turn this into an error by
-  // uncommenting the next line.
-  // llvm_unreachable("Kind of value not handled yet.");
-  return nullptr;
+  llvm_unreachable("Kind of value not handled yet.");
+}
+
+Value *LongVectorLoweringPass::visitConstant(Constant &Cst) {
+  auto *EquivalentTy = getEquivalentType(Cst.getType());
+  if (EquivalentTy == nullptr) {
+    return nullptr; // No replacement needed.
+  }
+
+  if (isa<UndefValue>(Cst)) {
+    return UndefValue::get(EquivalentTy);
+  }
+
+  if (auto *Vector = dyn_cast<ConstantDataVector>(&Cst)) {
```

Comment:
I think you are missing constant null vectors. They should be handled now. 

Separately, I assume it is left to future changes to handle constant expressions?

Reply:
I'm happy to track missing functionality in #613. To be clear, I'm totally ok with this coming together in small steps. These questions are more to help to me scope out the extent of a given PR.

Reply:
FYI, I've updated the description in #613.